ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java

```diff
@@ -0,0 +1,259 @@
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(metrics.getBlobSecurityProcessRequestMetrics, result -> {
+        String blobIdStr = requestPath.getOperationOrBlobId(false);
+        idConverter.convert(restRequest, blobIdStr, idConverterCallback());
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
+     */
+    private Callback<String> idConverterCallback() {
+      return buildCallback(metrics.getBlobIdConversionMetrics, convertedBlobId -> {
+        BlobId blobId = FrontendUtils.getBlobIdFromString(convertedBlobId, clusterMap);
+        if (restRequest.getArgs().get(TARGET_ACCOUNT_KEY) == null) {
+          // Inject account and container when they are missing from the rest request.
+          accountAndContainerInjector.injectTargetAccountAndContainerFromBlobId(blobId, restRequest, metricsGroup);
+        }
+        securityService.postProcessRequest(restRequest, securityPostProcessRequestCallback(blobId));
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#postProcessRequest} finishes, call {@link Router#getBlob} to dget
+     * the blob info from the storage layer.
+     * @param blobId the {@link BlobId} to get info
+     * @return a {@link Callback} to be used with {@link SecurityService#postProcessRequest}.
+     */
+    private Callback<Void> securityPostProcessRequestCallback(BlobId blobId) {
+      return buildCallback(metrics.getSecurityPostProcessRequestMetrics, result -> {
+        // inject encryption metrics if need be
+        if (BlobId.isEncrypted(blobId.getID())) {
+          RestRequestMetrics restRequestMetrics = metricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), true);
+          restRequest.getMetricsTracker().injectMetrics(restRequestMetrics);
+        }
+        if (subResource == null) {
+          router.getBlob(blobId.getID(), options, routerCallback(),
+              QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, true));
+        } else {
+          switch (subResource) {
+            case BlobInfo:
+            case UserMetadata:
+            case Segment:
+              router.getBlob(blobId.getID(), options, routerCallback(),
+                  QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, true));
+              break;
+            case Replicas:
+              finalCallback.onCompletion(getReplicasHandler.getReplicas(blobId.getID(), restResponseChannel), null);
+              break;
+          }
+        }
+      }, restRequest.getUri(), LOGGER, finalCallback);
```

Comment:
Looks like if we are still [calculating](https://github.com/linkedin/ambry/blob/master/ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java#L949) charges incurred for the GET request on router exceptions. Would we need same here too?

Reply:
it's weird, we already call the processRequestCharges in securityService.processResponse, it seems like we are doing duplicate work here.

Reply:
I am just thinking that securityService.processResponse might not be called if there is a router exception since we directly invoke final callback. But not sure if we need to calculate request charges on failures. Our current code seems to be doing that.